compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -431,45 +741,204 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+            }
+            // Do not track unions.
+            AggregateKind::Adt(_, _, _, _, Some(_)) => return None,
+        };
+
+        let fields: Option<Vec<_>> = fields
+            .iter_mut()
+            .map(|op| self.simplify_operand(op, location).or_else(|| self.new_opaque()))
+            .collect();
+        let fields = fields?;
+
+        if let AggregateTy::Array = ty && fields.len() > 4 {
+            let first = fields[0];
+            if fields.iter().all(|&v| v == first) {
+                let len = ty::Const::from_target_usize(self.tcx, fields.len().try_into().unwrap());
+                if let Some(const_) = self.try_as_constant(first) {
+                    *rvalue = Rvalue::Repeat(Operand::Constant(Box::new(const_)), len);
+                } else if let Some(local) = self.try_as_local(first, location) {
+                    *rvalue = Rvalue::Repeat(Operand::Copy(local.into()), len);
+                    self.reused_locals.insert(local);
+                }
+                return Some(Value::Repeat(first, len));
+            }
+        }
+
+        let value = Value::Aggregate(ty, variant_index, fields);
+        Some(value)
+    }
+}
+
+fn op_to_prop_const<'tcx>(
+    ecx: &mut InterpCx<'_, 'tcx, DummyMachine>,
+    op: &OpTy<'tcx>,
+) -> Option<ConstValue<'tcx>> {
+    // Do not attempt to propagate unsized locals.
+    if op.layout.is_unsized() {
+        return None;
+    }
+
+    // This constant is a ZST, just return an empty value.
+    if op.layout.is_zst() {
+        return Some(ConstValue::ZeroSized);
+    }
+
+    // Do not synthetize too large constants. Codegen will just memcpy them, which we'd like to avoid.
+    if !matches!(op.layout.abi, Abi::Scalar(..) | Abi::ScalarPair(..)) {
+        return None;
+    }
+
+    // If this constant has scalar ABI, return it as a `ConstValue::Scalar`.
+    if let Abi::Scalar(abi::Scalar::Initialized { .. }) = op.layout.abi
+        && let Ok(scalar) = ecx.read_scalar(op)
+        && scalar.try_to_int().is_ok()
+    {
+        return Some(ConstValue::Scalar(scalar));
+    }
+
+    // If this constant is a projection of another, we can return it directly.
+    if let Either::Left(mplace) = op.as_mplace_or_imm()
+        && let MemPlaceMeta::None = mplace.meta()
+    {
+        let (size, align) = ecx.size_and_align_of_mplace(&mplace).ok()??;
+
+        let alloc_ref = ecx.get_ptr_alloc(mplace.ptr(), size, align).ok()??;
+        // Do not try interning a value that contains provenance.
+        if alloc_ref.has_provenance() {
+            return None;
+        }
+
+        intern_const_alloc_for_constprop(ecx, &mplace).ok()?;
+        let pointer = mplace.ptr().into_pointer_or_addr().ok()?;
+        let (alloc_id, offset) = pointer.into_parts();
+        match ecx.tcx.global_alloc(alloc_id) {
+            GlobalAlloc::Memory(_) => return Some(ConstValue::Indirect { alloc_id, offset }),
+            // Fallthrough to copying the data.
+            _ => {}
+        }
```

Comment:
So you are calling `intern_const_alloc_for_constprop` and then you are double-checking its work? Why?
```suggestion
        let pointer = mplace.ptr().into_pointer_or_addr().ok()?;
        let (alloc_id, offset) = pointer.into_parts();
        intern_const_alloc_for_constprop(ecx, alloc_id).ok()?;
        Some(ConstValue::Indirect { alloc_id, offset });
```
(also changes `intern_const_alloc_for_constprop` to take an AllocId, since the caller needs it anyway and it's all the function cares for)